### ISSUE TITLE
0.5.12 breaks tab opening keyword javascript bookmarklets in fx3.6

### ISSUE THREAD

**Issue by jsamuel** [RequestPolicy/requestpolicy#72]
```
imported trac ticket
created: 2010-01-24 09:15:39
reporter: al_9x
```

Here is a example of such a bookmarklet, it performs a search in different engines in separate tabs
{{{
javascript:open('http://www.google.com/search?q=%s','_self');open('http://www.bing.com/search?q=%s');open('http://search.yahoo.com/search?p=%s');open('http://www.altavista.com/web/results?q=%s');open('http://www.ask.com/web?q=%s');
}}}

When RP is enabled, only the first url loads.  Verified on a new fx3.6 profile on xp32 with no other extensions.

**Comment by jsamuel** [comment]
```
imported trac comment
created: 2010-01-24 11:17:35
author: justin
```

Thanks for reporting this. However, I don't think this can be fixed. I also don't believe this works with any version of Firefox that !RequestPolicy supports. I've tried just now with 0.5.8 and Firefox 3.0.17 and it always fails, but the way it fails can vary by which page one is on (sometimes blank tabs open, sometimes Firefox blocks them as popups -- I'm guessing the latter might be due to a page I tested on possibly having redefined the name "open").

The problem is that, as far as I know, the bookmarklet will execute like javascript on the current tab you're on. With the information that !RequestPolicy has access to when deciding when to allow requests (specifically, nsIContentPolicy::shouldLoad() and any other data !RequestPolicy has collected ahead of time), there's no distinction between the tabs opened by this bookmarklet and tabs opened by the same javascript if it existed in the page you're on.

I think the way to solve this is to not try to perform this action through a bookmarklet and instead to use more privileged code to open these tabs. Basically, I think you'll need an extension. Maybe there's already an extension to provide this, possibly https://addons.mozilla.org/en-US/firefox/addon/1271, though I haven't looked too closely at that.

Anyways, please let me know if you are aware of any workarounds for this. Otherwise, I'm inclined to say that because of the way bookmarklets work, there's no great solution to this. The only workaround I could think to try wasn't successful: I changed the bookmarklet to first change the current tab's location to some trusted domain that's harmless for requests in and out  and is fully whitelisted as both an origin and destination (what that would be depend on the person), and then had it call open(). However, the bookmarklet stops executing once the page changes, it seems.

I do wish there was a way to solve this, because I imagine some people have a pretty strong need for bookmarklets that open() urls on other sites.

My final suggestion is to work around this by creating a page of your own that takes a request variable or information in the fragment of the uri and the javascript in that page opens the additional tabs. You'd need this page hosted somewhere as it wouldn't work to have it be a file:// uri. Let me know if you want to pursue that and need a hand with the code.

**Comment by jsamuel** [comment]
```
imported trac comment
created: 2010-01-24 12:16:11
author: al_9x
```

Replying to [comment:1 justin]:

> sometimes Firefox blocks them as popups

This bookmarklet functionality has been broken since fx3 (separate from RP) https://bugzilla.mozilla.org/show_bug.cgi?id=460880
in precisely the way you indicated, fx blocking them, but only when the places database is over a certain size.  If you try it in a new profile (small places db) with fx3-3.5.*, it should work.  However, it has been fixed (indirectly, not through my bug) in 3.6 and is working reliably (without RP). So as not to get mixed up with two separate problems, let's forget about fx versions < 3.6.

I use this functionality extensively and didn't upgrade to 3.0 and 3.5 primarily because of it.  It also indicated that there was something really screwy in fx3-3.5 for it to fail in such a bizarre way (from the size of the places db)

If I understood you correctly, RP is blocking these as unauthorized third party sub-requests of a previous navigation.  But since they are a result of an explicit user invocation I thought they would be treated as new first party requests.

Does fx give no clue that these requests are coming from a bookmark invocation?

I did just think of a workaround, that I would like to run by you.  I execute these bookmarklets primarily form a blank page, so adding about:blank to origins seems to work.  Are there any side effects to having about:blank in origins?  For example, would an iframe that starts out blank but is later given a source not be blocked? or anything along those lines?

**Comment by jsamuel** [comment]
```
imported trac comment
created: 2010-01-24 12:50:05
author: justin
```

I think using about:blank as a trusted origin would be a reasonable workaround if your paranoia level isn't high (mine is on the high side). Allowing the origin about:blank has been found to be a workaround in other cases, such as #18.

I'm not aware at the moment of any reliable method of making requests originate from about:blank, but I also wouldn't be willing to say it's not possible. For example, as a starting place to trying to run code from about:blank I see the following which could be useful: http://lcamtuf.coredump.cx/ffblank/ --- This is something I should really look into further so I can give a better answer regarding about:blank as an origin, because it's clearly something that will keep coming up.

Just to clarify what I was trying to explain before but probably am too tired to explain clearly: From what I understand and have seen, bookmarklets are just ways of injecting javascript into the current page you're looking at. They don't have any privilege that javascript on that page doesn't have, and everything they do looks like it came from javascript in that page. (I can't say for certain there isn't some difference other than the ability to use placeholders that can be used with the invocation name given to the bookmarklet, but as far as I know this is the case.)

It would easily solve your problem if there was a way to run bookmarket code in chrome's context (that is, fully privileged to do anything firefox can), but the potential for security problems there is quite high if people aren't careful about the bookmarklets they use. Maybe such functionality already exists with Greasemonkey.
